evm/src/keccak/keccak_stark.rs

```diff
@@ -1,60 +1,522 @@
+};
+use crate::keccak::registers::{
+    reg_a, reg_a_prime, reg_a_prime_prime, reg_a_prime_prime_0_0_bit,
+    reg_a_prime_prime_prime, reg_b, reg_c, reg_c_partial, reg_step, NUM_REGISTERS,
+};
+use crate::keccak::round_flags::{eval_round_flags, eval_round_flags_recursively};
 use crate::stark::Stark;
+use crate::util::trace_rows_to_poly_values;
 use crate::vars::{StarkEvaluationTargets, StarkEvaluationVars};
 
+/// Number of rounds in a Keccak permutation.
+pub(crate) const NUM_ROUNDS: usize = 24;
+
+/// Number of 64-bit limbs in a preimage of the Keccak permutation.
+pub(crate) const INPUT_LIMBS: usize = 25;
+
+pub(crate) const NUM_PUBLIC_INPUTS: usize = 0;
+
 #[derive(Copy, Clone)]
 pub struct KeccakStark<F, const D: usize> {
     pub(crate) f: PhantomData<F>,
 }
 
+impl<F: RichField + Extendable<D>, const D: usize> KeccakStark<F, D> {
+    /// Generate the rows of the trace. Note that this does not generate the permuted columns used
+    /// in our lookup arguments, as those are computed after transposing to column-wise form.
+    pub(crate) fn generate_trace_rows(
+        &self,
+        inputs: Vec<[u64; INPUT_LIMBS]>,
+    ) -> Vec<[F; NUM_REGISTERS]> {
+        let num_rows = (inputs.len() * NUM_ROUNDS).next_power_of_two();
+        info!("{} rows", num_rows);
+        let mut rows = Vec::with_capacity(num_rows);
+        for input in inputs.iter().take(1) {
+            rows.extend(self.generate_trace_rows_for_perm(*input));
+        }
+
+        // Pad rows to power of two.
+        for i in rows.len()..num_rows {
+            let mut row = [F::ZERO; NUM_REGISTERS];
+            self.copy_output_to_input(rows[i - 1], &mut row);
+            self.generate_trace_rows_for_round(&mut row, i % NUM_ROUNDS);
+            rows.push(row);
+        }
+
+        rows
+    }
+
+    fn generate_trace_rows_for_perm(&self, input: [u64; INPUT_LIMBS]) -> Vec<[F; NUM_REGISTERS]> {
+        let mut rows = vec![[F::ZERO; NUM_REGISTERS]; NUM_ROUNDS];
+
+        for x in 0..5 {
+            for y in 0..5 {
+                let input_xy = input[x * 5 + y];
+                for z in 0..64 {
+                    rows[0][reg_a(x, y, z)] = F::from_canonical_u64((input_xy >> z) & 1);
+                }
+            }
+        }
+
+        self.generate_trace_rows_for_round(&mut rows[0], 0);
+        for round in 1..24 {
+            self.copy_output_to_input(rows[round - 1], &mut rows[round]);
+            self.generate_trace_rows_for_round(&mut rows[round], round);
+        }
+
+        rows
+    }
+
+    fn copy_output_to_input(
+        &self,
+        prev_row: [F; NUM_REGISTERS],
+        next_row: &mut [F; NUM_REGISTERS],
+    ) {
+        for x in 0..5 {
+            for y in 0..5 {
+                let cur_lo = prev_row[reg_a_prime_prime_prime(x, y)];
+                let cur_hi = prev_row[reg_a_prime_prime_prime(x, y) + 1];
+                let cur_u64 = cur_lo.to_canonical_u64() + (1 << 32) * cur_hi.to_canonical_u64();
```

Comment:
nit: I think it's slightly more idiomatic to combine them with logic operators like `cur_lo.to_canonical_u64() | (cur_hi.to_canonical_u64() << 32)`

Reply:
And optionally `*acc % 2` -> `*acc & 1`, `*acc /= 2` -> `*acc >>= 1`, but I guess it's just a matter of preference, fine to keep as is if you prefer

Reply:
Ah yes that looks better!

Reply:
Hm in that case should I also replace `val % (1 << 32)` with `val & (1 << 32 - 1)`?

Reply:
Hm up to you. Another option is `val as u32 as u64`

Reply:
Oh interesting. I think I'll just go with the bit-logic one